### ISSUE TITLE
Disallow implicit Any in types

### ISSUE THREAD

**Issue by rgabbard** [isi-vista/vistautils#86]
_Originally posted by @berquist in https://github.com/isi-vista/vistautils/pull/84_

**Comment by jamart28** [comment]
@gabbard Is this just wanting mypy to disallow implicit `Any`? If so does this flag work?
```
  --disallow-any-generics   disallow usage of generic types that do not
                            specify explicit type parameters
```

**Comment by gabbard** [comment]
@jamart28 : 

I think we want:

```
--disallow-any-generics ----disallow-any-expr --disallow-any-unimported
```

I would recommend adding them one-by-one so you can identify which issues come from which source.

**Comment by jamart28** [comment]
So far adding just `--disallow-any-generics` causes all instances where we don't specify a size for `Tuple` errors out in mypy. One such fix is to replace such `Tuple`s which seem to be variable or at least large sizes but all one type with `Sequence`. Would this be the preferred solution?

**Comment by gabbard** [comment]
@jamart28 : Can you show me an example of such an error?

**Comment by jamart28** [comment]
![image](https://user-images.githubusercontent.com/44595288/74265752-fdb11f80-4cfa-11ea-934b-17e24ee6d1f5.png)

preconditions.py:8:
```
_ClassInfo = Union[type, Tuple[Union[type, Tuple], ...]]
```
@gabbard

**Comment by gabbard** [comment]
@jamart28 : That particular example is weird and can be `type: ignore`d. Can you show me another example (or several)?

**Comment by jamart28** [comment]
Yes let me find one more straightforward of the same type

**Comment by jamart28** [comment]
Looking back through it seems like these situations may actually have to do with the inside typings (for the one above the inside `Tuple` has no parameter)
As such this works:
```
_ClassInfo = Union[type, Tuple[Union[type, Tuple[Any, ...]], ...]]
```
Is this preferable or is something more specific needed? Furthermore, should I just replace all implicit Any's with an explicit Any or attempt to classify something more specific? @gabbard

**Comment by gabbard** [comment]
@jamart28 : In the `_ClassInfo` situation, I think this is fine.  I would need to see more examples to comment more generally.

**Comment by jamart28** [comment]
@gabbard: Of the situation with the dots or all the errors after updating mypy?

**Comment by gabbard** [comment]
All the errors after updating mypy

**Comment by jamart28** [comment]
![image](https://user-images.githubusercontent.com/44595288/74270319-ef670180-4d02-11ea-9c14-fbe02215107f.png)

This is just with the first flag above and are all caused by a form of having a generic with no typing after it for what is expected for the generic. In my opinion, the easiest way to show all these would be for me to fix them and then open a draft pull request showing the diffs. I was planning on doing this with each flag above. Would this be acceptable or is a different course of action preferred?

**Comment by gabbard** [comment]
That is a good approach. If any don't have obvious fixes, link me to the relevant lines of code and I will comment.

**Comment by jamart28** [comment]
@gabbard: Is there any particular reason we type some variables and not others? I'm specifically referring to a situation like lines 232, 380, and 600 within `vistautils/io_utils.py`. Since all of these are assigned immediately and not reassigned anywhere in their function, wouldn't it be favorable to a long typing to have no typing (almost simulating an assumed typing from something like kotlin)? Also considering that's basically what's done with the second two lines.

**Comment by gabbard** [comment]
@jamart28 : In some cases we may have underestimated the type inference, so it is possible some type annotations could be removed without harm.

**Comment by jamart28** [comment]
@gabbard: So for a quick clarification to hold over until I look at mypy a little closer, does mypy do type inference on variables if possible?

**Comment by gabbard** [comment]
@jamart28 : `mypy` attempts to do some degree of type inference, though what it can do is usually more limited than e.g. Kotlin.  See https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html

**Comment by jamart28** [comment]
@gabbard: Do you see anything in the line below that would match this error: `vistautils/key_value.py:169: error: Missing type parameters for generic type`? https://github.com/isi-vista/vistautils/blob/e1870563b7283ab189878c7c600cb3639f9c8d2d/vistautils/key_value.py#L169

**Comment by gabbard** [comment]
Odd - my guess would be that `AbstractContextManager` is generic, though I don't see it documented as such?

Can you see if you can reproduce this error with a minimal example of a generic class deriving from `AbstractContextManager`?

**Comment by gabbard** [comment]
Aha, it looks like we should be using [`typing.ContextManager`](https://docs.python.org/3/library/typing.html#typing.ContextManager), which is generic, though it doesn't document exactly what the type of the argument is.  I believe it should be:

```python
 class KeyValueLinearSource(Generic[K, V], ContextManager["KeyValueLinearSource"], metaclass=ABCMeta): 

```

**Comment by jamart28** [comment]
I can try that.

**Comment by jamart28** [comment]
Even with that the issue persists and throws all the following errors with it: 
![image](https://user-images.githubusercontent.com/44595288/74279819-905db880-4d13-11ea-81c9-985a2ea6aa99.png)

**Comment by jamart28** [comment]
Ok. Doing
```
class KeyValueLinearSource(Generic[K, V], AbstractContextManager[KeyValueLinearSource[Any, Any]], metaclass=ABCMeta):
```
gets rid of the `Missing type parameters for generic type` but again gives all the above errors.

**Comment by jamart28** [comment]
As such one fix is: 
```
class KeyValueLinearSource(Generic[K, V], AbstractContextManager[Any], metaclass=ABCMeta):
```
Let me know if this is acceptable.

**Comment by gabbard** [comment]
What if you use `K,V` as the args to the inner instance of the `KeyValueLinearSource` type?

**Comment by jamart28** [comment]
We just get more specific errors sadly: 
![image](https://user-images.githubusercontent.com/44595288/74280546-a4ee8080-4d14-11ea-8be0-4712350120e3.png)

**Comment by gabbard** [comment]
@jamart28 : poke me about this tomorrow and I can help you more. In the meantime, work on other classes of errors.

**Comment by jamart28** [comment]
I can do that. I'll move over to something else for today as the next issue I will most likely need help with as well since I have the cause of the issue narrowed down but not the why or the fix.

**Comment by jamart28** [comment]
The following error occurs when running with just the `--disallow-any-generics` flag for now: 
![image](https://user-images.githubusercontent.com/44595288/74853289-a1bb4c00-5335-11ea-9c42-9e35e5651259.png) I've traced this to having an issue with the import of `immutabledict` from `immutablecollections`. As such one fix would be to trace the error through the import and fix it there, but another is to possibly add a flag to ignore imports (`--follow-imports skip`) however this seems to throw another error so further research would be needed. Most likely the best solution would be to trace the error through the import and fix it there. If that's the solution wanted I can set up an issue and work it on that repo.

**Comment by jamart28** [comment]
Found the solution for the first problem (the one @gabbard was originally helping me with). Adding an `Any` causes an error on actually running the program since the `AbstractContextManager` is not subscriptable, but `mypy` wants it to have a typing. As such, the possible solutions are to type ignore (the better in my opinion) or figure out what typing mypy will accept and handling the situation based on type checking like [this](https://github.com/python/mypy/issues/5264).

**Comment by LichMaster98** [comment]
@jamart28 Correct, we should go try to make this fix in the `immutablecollections` codebase. 

I think I'm going to defer to @gabbard on the route for the problem you were originally discussing with him.

**Comment by jamart28** [comment]
@LichMaster98 The next two flags suggested above produce the following results. I would like to know if these are intended results and thus should be pursued for a fix or if they should not be added:
`--disallow-any-expr`: This disallows all `Any`'s. As such it targets explicit `Any`'s as well, not allowing `Any` to be used in any capacity.
`--disallow-any-unimported`: This one disallows `Any`'s assumed from a type that was imported but not followed by mypy thus causing mypy to assume `Any`. Realistically the only errors this causes has to do with the following import not being followed in range.py: https://github.com/isi-vista/vistautils/blob/0b0be9d842fb4d547b3e020a6b5687cc6105ca72/vistautils/range.py#L26

**Comment by LichMaster98** [comment]
@jamart28 
1) I'm unsure if we want to eliminate the ability to Explicitly use Any

2) We might be able to implement the second while just explicitly ignoring the issue on `SortedDict`.

**Comment by jamart28** [comment]
1. That was my though process but I wanted to check as it was in the list of flags to check
2. We certainly could I just was unsure if it was necessary. I can definitely throw some `# type: ignore`s in and enable that flag.
